enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,447 @@
+to ensure that the API on the control plane is fully functional before the switch.
+
+The Ingress VIPs will be created on the control plane nodes where the Ingress Controller reports healthy.
+
+FRR will be deployed as a [static pod](https://kubernetes.io/docs/tasks/administer-cluster/static-pod/) and the assets will be rendered by the Machine Config Operator.
+No consideration needs to be made when replacing control plane machines for the stability of the system, because multiple BGP sessions to the leaf can be done at the same time.
+
+#### Init containers
+
+- `copy-frr-static-configs` will provide static files to put in /etc/frr, like how we configure daemons and vtysh.conf (required for healthcheck).
+- `verify-api-int-resolvable` on the control plane will ensure that the internal VIP can be resolved before creating it.
+- `render-config-frr` will use baremetal-runtimecfg to render `frr.conf`. Below we provide details on this file.
+
+#### FRR container
+
+This container will handle the start, stop and reload of BGP and BFD daemons in FRR.
+
+- It'll wait for the `frr.conf` to exist (with a timeout) and then start daemons if it exists
+- It'll stop daemons if it gets a `stop` message via a socat handler running with baremetal-runtimecfg, like it's already done for HAproxy & Keepalived (and re-using the same conditions if we need to stop it for
+  feature parity).
+- It'll reload daemons if it gets a `reload` message from the socat handler, if configuration on-disk was changed with the one currently running.
+
+FRR's configuration will be opinionated but flexible enough to meet customer needs. Each Failure Domain will have its BGP configuration:
+- A list of BGP peers (usually their ToR routers) IP
+- Each peer would have an IP address, a BGP password and an ASN
+
+The customers will not be able to customize FRR until we find some real limitations. BGP and BFD daemons will be enabled by `copy-frr-static-configs` container.
+
+#### VIPs management
+
+This container will handle the creation and deletion of the VIPs.
+Here's a programmatic way to describe how VIPs will be managed:
+
+```bash
+while :; do
+    if check_health "api"; then
+      echo "Healthcheck for API OK"
+      for vip in $API_VIPS; do
+        create_vip $vip
+        create_vip_routing $vip
+      done
+    else
+      echo "Healthcheck for API not OK"
+      for vip in $API_VIPS; do
+        delete_vip $vip
+      done
+    fi
+
+    if check_health "ingress"; then
+      echo "Healthcheck for Ingress OK"
+      for vip in $INGRESS_VIPS; do
+        create_vip $vip
+        create_vip_routing $vip
+      done
+    else
+      echo "Healthcheck for Ingress not OK"
+      for vip in $INGRESS_VIPS; do
+        delete_vip $vip
+      done
+    fi
+    sleep 2
+done
+```
+
+##### Health check
+
+When checking for the health of a service (API and Ingress), we'll run a `curl` against the local service (API or Ingress) and check the response code.
```

Comment:
Ideally users will want to ensure that control plane membership changes like this have a minimal effect on existing connections. Naively this could mean that if an API node goes out, rehashing by the upstream router could result in connection resets on almost all existing connections. Can something like consistent hashing be useful with only 3 nodes? Either way, this is likely an important documentation issue.

Reply:
I'm not sure what to say about this in the RFE, if you can help me on this one. I understand what you're saying, but I guess we also have the same issue with Keepalived today, so trying to see what needs to be said in this RFE, that matters for BGP.